File: java/examples/src/main/java/org/pactman/nonprofitcheckplus/examples/cases/Ex08AutomaticRevocation.java
package org.pactman.nonprofitcheckplus.examples.cases;

import org.pactman.nonprofitcheckplus.Sources;
import org.pactman.nonprofitcheckplus.devtools.FixtureEins;
import org.pactman.nonprofitcheckplus.examples.support.Example;
import org.pactman.nonprofitcheckplus.examples.support.ExampleContext;
import org.pactman.nonprofitcheckplus.examples.support.IrsCodes;
import org.pactman.nonprofitcheckplus.examples.support.Output;
import org.pactman.nonprofitcheckplus.models.AroeSource;
import org.pactman.nonprofitcheckplus.models.Nonprofit;

/**
 * EX-08 — The Automatic Revocation of Exemption list.
 *
 * <p>An organization that fails to file for three consecutive years loses its
 * exemption automatically. The revocation fields say so, and they say it
 * independently of what the BMF and Publication 78 report.
 */
public final class Ex08AutomaticRevocation implements Example {

    @Override
    public String id() {
        return "ex-08";
    }

    @Override
    public String title() {
        return "Automatic revocation, read across every source";
    }

    @Override
    public void run(String[] args) {
        try (ExampleContext context = ExampleContext.withFixtures()) {
            for (String ein : new String[] {FixtureEins.PUBLIC_CHARITY, FixtureEins.REVOKED}) {
                Nonprofit nonprofit = context.client().nonprofits().check(ein).getNonprofit();
                AroeSource aroe = Sources.aroe(nonprofit);

                Output.heading(nonprofit.getOrganizationName() + " (" + ein + ")");

                if (aroe == null) {
                    Output.bullet("The API returned no revocation data at all.");
                    continue;
                }

                Output.field("revocation_code", aroe.getRevocationCode());
                Output.field(
                        "  meaning", IrsCodes.describeRevocationCode(aroe.getRevocationCode()));
                Output.field("revocation_date", aroe.getRevocationDate());
                Output.field("reinstatement_date", aroe.getReinstatementDate());

                // The other sources agree here, and that is worth showing: a
                // revocation is visible in three places, and a workflow that
                // reads only one of them is one API change from missing it.
                Output.field("pub78_verified", nonprofit.getPub78Verified());
                Output.field("bmf_status", nonprofit.getBmfStatus());
                Output.field("exempt_status_code", nonprofit.getExemptStatusCode());
                Output.field(
                        "  meaning",
                        IrsCodes.describeExemptStatus(nonprofit.getExemptStatusCode()));
            }
        }

        Output.note("A revocation date with no reinstatement date is the state worth acting on.\n"
                + "The next example shows why the reinstatement date is not optional reading.");
    }
}

File: ruby/examples/ex_08_automatic_revocation.rb
# frozen_string_literal: true

# EX-08 — Automatic revocation detected.
#
# An organization that fails to file for three consecutive years has its
# exemption revoked automatically and appears in the IRS Automatic Revocation of
# Exemption (AROE) data. The API reports that with `revocation_code` and
# `revocation_date`.
#
# This example flags the record and preserves the source fields verbatim. It does
# not decide the outcome — blocking, holding, or reviewing is your policy.
#
# Run:  PACTMAN_API_KEY=... bundle exec ruby examples/ex_08_automatic_revocation.rb

require "time"
require_relative "lib/fixture_api"
require_relative "lib/print"

NCP = Pactman::NonprofitCheckPlus
include Print

# The application's policy, in one place, expressed against source fields.
POLICY = {
  on_revoked_without_reinstatement: "block",
  on_revoked_with_reinstatement: "manual_review",
  on_no_revocation_data: "continue"
}.freeze

Assessment = Data.define(:action, :reason, :aroe)

def filled?(value)
  !value.nil? && value != ""
end

def assess_revocation(nonprofit)
  aroe = NCP::Sources.aroe(nonprofit)

  return Assessment.new(POLICY[:on_no_revocation_data], "No revocation fields were returned.", aroe) if aroe.nil?

  unless filled?(aroe.revocation_code) || filled?(aroe.revocation_date)
    return Assessment.new(POLICY[:on_no_revocation_data], "Revocation fields were returned and are empty.", aroe)
  end

  if filled?(aroe.reinstatement_date)
    Assessment.new(POLICY[:on_revoked_with_reinstatement], "Revoked, with a reinstatement date present — see ex_09.",
                   aroe)
  else
    Assessment.new(POLICY[:on_revoked_without_reinstatement],
                   "Appears in the Automatic Revocation data with no reinstatement.", aroe)
  end
end

FixtureApi.with_fixture_api do |client|
  [FixtureApi::EINS[:revoked], FixtureApi::EINS[:public_charity]].each do |ein|
    result = client.nonprofits.check(ein)
    nonprofit = result.nonprofit

    if nonprofit.nil?
      puts "No record for #{ein}."
      next
    end

    assessment = assess_revocation(nonprofit)
    bmf = NCP::Sources.bmf(nonprofit)

    heading "#{nonprofit.organization_name} (#{nonprofit.ein})"
    field "revocation_code", returned(assessment.aroe, :revocation_code)
    field "revocation_date", returned(assessment.aroe, :revocation_date)
    field "reinstatement_date", returned(assessment.aroe, :reinstatement_date)

    # Revocation shows up in the other sources too. Capture what each one said,
    # rather than letting one field speak for all of them.
    field "bmf_status", returned(bmf, :status)
    field "pub78_verified", returned(NCP::Sources.pub78(nonprofit), :verified)
    field "exempt_status_code", returned(bmf, :exempt_status_code)

    field "policy action", assessment.action
    field "reason", assessment.reason

    # What you keep is what you can explain later. Store the source fields, the
    # request identifier, and the time you looked — not just the verdict.
    audit_record = {
      ein: nonprofit.ein,
      checked_at: Time.now.utc.iso8601,
      request_id: result.request_id,
      action: assessment.action,
      source_findings: nonprofit.to_h.slice(
        "revocation_code", "revocation_date", "reinstatement_date", "bmf_status", "pub78_verified"
      )
    }

    puts "\n  audit record:"
    json_block audit_record
  end
end

note "The SDK reports what the AROE data says. It does not decide whether a revoked\n" \
     "organization may receive a donation, a grant, or a payout — that is a legal and\n" \
     "compliance determination your application owns."
